### PR TITLE
Exception handling

### PR DESCRIPTION
**Pull request by matthiashimmelmann**
+ Pinned vertices moved to rigidity matrix
+ Assertions turned into Exceptions

### CODE REVIEW

File: pyrigi/framework.py [hunk 1]
```diff
@@ -62,31 +57,27 @@ def __init__(self,
                  realization: Dict[Vertex, Point] = {},
                  pinned_vertices: Dict[Vertex, List[int]] = {},
                  dim: int = 2) -> None:
-        assert isinstance(graph, Graph)
+        if not isinstance(graph, Graph): 
+            raise TypeError("Graph is not of the correct Type!")
```
Comment (MatteoGallet) [inline pyrigi/framework.py:61]: Maybe better "The parameter `graph has to be an instance of class Graph"?

Comment (MatteoGallet) [inline pyrigi/framework.py:61]: Maybe  "The  `graph` has to be an instance of class Graph" ?

File: pyrigi/framework.py [hunk 2]
```diff
@@ -62,31 +57,27 @@ def __init__(self,
...
+        if not len(realization.keys()) == len(graph.vertices()): 
+            raise AttributeError("Realization and Graph need to have the same number of entries!")
```
Comment (MatteoGallet) [inline pyrigi/framework.py:63]: Maybe "The length of `realization` has to be equal to the number of vertices of  `graph`"?

File: pyrigi/framework.py [hunk 3]
```diff
@@ -62,31 +57,27 @@ def __init__(self,
...
             dimension = len(list(realization.values())[0])
 
-        assert len(realization.keys()) == len(graph.vertices())
         for v in graph.vertices():
-            assert v in realization
-            assert len(realization[v]) == dimension
-            if not v in pinned_vertices:
-                pinned_vertices[v] = []
+            if not v in realization:
+                raise KeyError("Realization needs to contain the same vertices as Graph!")
```
Comment (MatteoGallet) [inline pyrigi/framework.py:73]: Maybe f"Vertex {v} is not a key of `realization`" ?

File: pyrigi/framework.py [hunk 4]
```diff
@@ -62,31 +57,27 @@ def __init__(self,
...
+            if not len(realization[v]) == dimension:
+                raise ValueError("An entry of the realization does not have the correct dimension!")
```
Comment (MatteoGallet) [inline pyrigi/framework.py:75]: Maybe f"The element {realization[v]} does not have the correct dimension" ?

File: pyrigi/framework.py [hunk 5]
```diff
@@ -104,15 +95,18 @@ def add_vertex(self, point: Point, vertex: Vertex = None) -> None:
             while candidate in self._graph.vertices():
                 candidate += 1
             vertex = candidate
-        assert vertex not in self._graph.vertices()
+
+        if vertex in self._graph.vertices():
+            raise AttributeError(str("Vertex is already contained in the graph's vertices!"))
```
Comment (MatteoGallet) [inline pyrigi/framework.py:100]: Maybe "The given `vertex` is already a vertex of `graph`" ?

File: pyrigi/framework.py [hunk 6]
```diff
@@ -104,15 +95,18 @@ def add_vertex(self, point: Point, vertex: Vertex = None) -> None:
...
         self._realization[vertex] = Matrix(point)
         self._graph.add_node(vertex)
-        self._pinned_vertices[vertex] = []
 
     def add_vertices(self,
                      points: List[Point],
                      vertices: List[Vertex] = []) -> None:
-        assert (len(points) == len(vertices) or not vertices)
+        if not (len(points) == len(vertices) or not vertices):
+            raise ValueError("The vertex list does not have the correct length!")
```
Comment (MatteoGallet) [inline pyrigi/framework.py:109]: Maybe "The lists `points` and `vertices` must have the same length" ?

File: pyrigi/framework.py [hunk 7]
```diff
@@ -121,8 +115,10 @@ def add_vertices(self,
                 self.add_vertex(p, v)
 
     def add_edge(self, edge: Edge) -> None:
-        assert (len(edge)) == 2
-        assert (edge[0] in self._graph.nodes and edge[1] in self._graph.nodes)
+        if not (len(edge)) == 2:
+            raise TypeError("Edge does not have the correct length!")
```
Comment (MatteoGallet) [inline pyrigi/framework.py:119]: Maybe "The list `edge` must have length 2" ?

File: pyrigi/framework.py [hunk 8]
```diff
@@ -121,8 +115,10 @@ def add_vertices(self,
...
+        if not (edge[0] in self._graph.nodes and edge[1] in self._graph.nodes):
+            raise ValueError("The edge's end points are not contained in the graph's vertices!")
```
Comment (MatteoGallet) [inline pyrigi/framework.py:121]: Maybe "The elements of `edge` must be vertices of the underlying graph of the framework" ?

File: pyrigi/framework.py [hunk 9]
```diff
@@ -144,11 +140,10 @@ def graph(self) -> Graph:
         """
         return self.underlying_graph()
 
-    def print(self) -> None:
+    def __str__(self) -> None:
```
Comment (MatteoGallet) [inline pyrigi/framework.py:143]: The method __str__ should return a single string and not print.

File: pyrigi/framework.py [hunk 10]
```diff
@@ -208,13 +203,17 @@ def realization(self) -> List[Point]:
 
     def set_realization(self, realization: Dict[Vertex, Point]) -> None:
         for v in self._graph.vertices():
-            assert v in realization
-            assert len(realization[v]) == self.dimension()
+            if not v in realization:
+                raise KeyError("Vertex is not contained in the given realization!")
```
Comment (MatteoGallet) [inline pyrigi/framework.py:207]: Maybe f"The vertex {v} is not a key of `realization`" ?

File: pyrigi/framework.py [hunk 11]
```diff
@@ -208,13 +203,17 @@ def realization(self) -> List[Point]:
...
+            if not len(realization[v]) == self.dimension():
+                raise AttributeError("The realization does not have the correct dimension at some vertex!")
```
Comment (MatteoGallet) [inline pyrigi/framework.py:209]: Maybe f"The element {realization[v]} does not have the correct dimension" ?

File: pyrigi/framework.py
Comment (MatteoGallet) [inline pyrigi/framework.py:1]: In general, maybe substitute AttributeError by ValueError?

File: pyrigi/framework.py [hunk 12]
```diff
@@ -248,38 +248,32 @@ def change_realization(self, subset_of_realization: Dict[Vertex, Point]):
...
         except TypeError as error:
             vertex_order = self._graph.vertices()
 
+        for v in vertex_order:
+            if not v in pinned_vertices:
+                pinned_vertices[v] = []
+        pinned_vertices = {v:pinned_vertices[v] for v in pinned_vertices.keys() if v in self._graph.vertices()}
+        for v in pinned_vertices:
+            if not v in self._graph.vertices():
+                raise KeyError("A vertex in pinned_vertices is not contained in the graph!")
```
Comment (MatteoGallet) [inline pyrigi/framework.py:273]: Maybe f"Vertex {v} in `pinned_vertices` is not a vertex of the graph" ?

File: pyrigi/framework.py [hunk 13]
```diff
@@ -248,38 +248,32 @@ def change_realization(self, subset_of_realization: Dict[Vertex, Point]):
...
+            if not len(pinned_vertices[v]) <= self.dim():
+                raise AttributeError("Too many coordinates provided for pinned_vertices!")
```
Comment (MatteoGallet) [inline pyrigi/framework.py:275]: Maybe f"The value {pinned_vertices[v]} is larger than the dimension" ?

File: pyrigi/framework.py [hunk 14]
```diff
@@ -307,35 +301,36 @@ def delta(u, v, w):
...
         r"""The complete graph is infinitesimally rigid in all dimensions. Thus, for computing the trivial
         flexes it suffices to compute all infinitesimal flexes of the complete graph."""
         Kn = nx.complete_graph(len(self._graph.vertices()))
         F_Kn = Framework(
             graph=Graph(Kn.edges),
             realization=self.realization(),
-            pinned_vertices=self.get_pinned_vertices(),
             dim=self.dim())
-        return F_Kn.infinitesimal_flexes(include_trivial=True)
+        return F_Kn.infinitesimal_flexes(pinned_vertices = pinned_vertices, include_trivial = True)
```
Comment (MatteoGallet) [inline pyrigi/framework.py:318]: Please use `pinned_vertices=pinned_vertices` and `include_trivial=True`.

**Review by MatteoGallet — COMMENTED** [review]
So far, I only looked at the strings in  the error raised, I still have to check the code.

File: pyrigi/framework.py (hunk 13 above)
Comment (matthiashimmelmann) [reply pyrigi/framework.py:275]: Ah, nice, I didn't know how String interpolation works in Python. This is obviously superior.

File: pyrigi/framework.py (hunk 5 above)
Comment (matthiashimmelmann) [reply pyrigi/framework.py:100]: I don't think there is a nice way to print the graph.

**Review by MatteoGallet — APPROVED** [review]

**Pull request merged**